### ISSUE TITLE
Mypy Errors

### ISSUE THREAD

**Issue by smessmer** [pytorch/pytorch#39192]
## 🐛 Bug

I'm seeing several errors running mypy, both on master and on the 1.5 release. Are we not checking mypy in CI?

Examples:
- on 1.5 release tag:
```
(pytorch) [messmer@devvm222.vll0 ~/pytorch] python -m mypy .
caffe2/python/layers/layers.py:247: error: misplaced type annotation
Found 1 error in 1 file (checked 980 source files)
```
- On master:
```
(pytorch) [messmer@devvm222.vll0 ~/pytorch] mypy
test/test_type_hints.py:13: error: Skipping analyzing 'mypy': found module but no type hints or library stubs  [import]
test/test_type_hints.py:13: note: See https://mypy.readthedocs.io/en/latest/running_mypy.html#missing-imports
Found 1 error in 1 file (checked 983 source files)
```
- On master, running on a single file (side note: why didn't those get reported above when running on the whole repo?)
```
(pytorch) [messmer@devvm222.vll0 ~/pytorch] mypy aten/src/ATen/native_parse.py
aten/src/ATen/native_parse.py:12: error: Incompatible import of "Loader" (imported name has type "Type[Loader]", local name has type "Type[CLoader]")  [misc]
aten/src/ATen/native_parse.py:27: error: Incompatible types in assignment (expression has type "List[str]", variable has type "str")  [assignment]
aten/src/ATen/native_parse.py:89: error: Item "None" of "Optional[Match[Any]]" has no attribute "group"  [union-attr]
aten/src/ATen/native_parse.py:93: error: Item "None" of "Optional[Match[Any]]" has no attribute "group"  [union-attr]
aten/src/ATen/native_parse.py:101: error: Item "None" of "Optional[Match[Any]]" has no attribute "group"  [union-attr]
aten/src/ATen/native_parse.py:110: error: Incompatible types in assignment (expression has type "bool", variable has type "Optional[str]")  [assignment]
aten/src/ATen/native_parse.py:112: error: Incompatible types in assignment (expression has type "bool", variable has type "Optional[str]")  [assignment]
aten/src/ATen/native_parse.py:140: error: Incompatible types in assignment (expression has type "int", variable has type "Optional[str]")  [assignment]
aten/src/ATen/native_parse.py:143: error: Incompatible types in assignment (expression has type "float", variable has type "Optional[str]")  [assignment]
aten/src/ATen/native_parse.py:151: error: Need type annotation for 'arguments' (hint: "arguments: List[<type>] = ...")  [var-annotated]
aten/src/ATen/native_parse.py:331: error: Need type annotation for 'arguments' (hint: "arguments: List[<type>] = ...")  [var-annotated]
aten/src/ATen/native_parse.py:401: error: Incompatible types in assignment (expression has type "bool", target has type "str")  [assignment]
Found 12 errors in 1 file (checked 1 source file)
```

cc @ezyang @rgommers

**Comment by rgommers** [comment]
Yes we are running mypy in CI. I'm getting:
```
Success: no issues found in 981 source files
```
right now.

> `mypy aten/src/ATen/native_parse.py`

That file isn't covered - see `files = ...` in `mypy.ini` in the root of the repo.

> `Skipping analyzing 'mypy': found module but no type hints or library stubs `

I think you're trying to use an installed PyTorch here. That isn't supported yet, we should complete coverage and ship a `py.typed` file (see [PEP 561](https://www.python.org/dev/peps/pep-0561/)) before that works. With a `setup.py develop` type build running `mypy` in the repo root should pass.

**Comment by rgommers** [comment]
tl;dr type annotation coverage is growing quickly, but it's still far from complete. gh-16574 is the tracking issue for it.

**Comment by smessmer** [comment]
Thanks. Seems I had several versions of pytorch installed. I now uninstalled all of them (call "pip uninstall torch" until "pip list" doesn't show it anymore), then did a "python setup.py develop", but I'm still seeing

```
(pytorch) [messmer@devvm222.vll0 ~/pytorch] python -m mypy
test/test_type_hints.py:13: error: Skipping analyzing 'mypy': found module but no type hints or library stubs  [import]
test/test_type_hints.py:13: note: See https://mypy.readthedocs.io/en/latest/running_mypy.html#missing-imports
Found 1 error in 1 file (checked 983 source files)
```

**Comment by smessmer** [comment]
Uninstalling mypy from conda and installing it with pip resolved the issue
